File: solutions/leetcode/3sum/main.cpp
class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if (nums.size() < 3) return {};

        unordered_map<int, int> num_count;
        vector<int> dedup_nums;
        for (int v : nums) {
            if (++num_count[v] == 1) dedup_nums.push_back(v);
        }

        sort(dedup_nums.begin(), dedup_nums.end());

        vector<vector<int>> result;
        for (int i = 0; i < dedup_nums.size(); ++i) {
            int l = i, r = dedup_nums.size() - 1;
            while (l <= r) {
                int a = dedup_nums[i];
                int b = dedup_nums[l];
                int c = dedup_nums[r];

                int sum = a + b + c;
                if (sum < 0) {
                    ++l;
                    continue;
                } else if (sum > 0) {
                    --r;
                    continue;
                }

                if ((i == l && l == r && num_count[a] >= 3) ||
                    (i == l && l != r && num_count[a] >= 2) ||
                    (i != l && l == r && num_count[b] >= 2) ||
                    (i != l && l != r)) {
                    result.push_back({a, b, c});
                }
                ++l;
            }
        }

        return result;
    }
};

class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        vector<vector<int>> result;

        for (int i = 0; i < nums.size(); ++i) {
            int l = i + 1, r = nums.size() - 1;
            while (l < r) {
                int s = nums[i] + nums[l] + nums[r];
                if (s < 0) {
                    ++l;
                    continue;
                }
                if (s > 0) {
                    --r;
                    continue;
                }

                result.push_back({nums[i], nums[l], nums[r]});
                while (l < r && nums[l] == nums[l + 1]) ++l;
                ++l;
            }

            while (i + 1 < nums.size() && nums[i] == nums[i + 1]) ++i;
        }

        return result;
    }
};

File: solutions/leetcode/3sum/main.py
class Solution:

    def twoSumSorted(self, nums: list[int], start: int,
                     target: int) -> list[list[int]]:
        l = start
        r = len(nums) - 1

        result = []

        while l < r:
            if nums[l] + nums[r] > target:
                r = r - 1
                continue

            if nums[l] + nums[r] == target:
                result.append([nums[l], nums[r]])

            l = l + 1
            while l < r and nums[l] == nums[l - 1]:
                l = l + 1

        return result

    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums = sorted(nums)

        result = []

        p = 0
        while p + 2 < len(nums):
            candidates = self.twoSumSorted(nums, p + 1, -nums[p])
            for l, r in candidates:
                result.append([nums[p], l, r])

            p = p + 1
            while p + 2 < len(nums) and nums[p] == nums[p - 1]:
                p = p + 1

        return result


class Solution:

    def twoSumSorted(self, nums: list[int], start: int,
                     target: int) -> list[list[int]]:
        l = start
        r = len(nums) - 1

        result = []

        while l < r:
            s = nums[l] + nums[r]
            if s < target:
                l += 1
            elif s > target:
                r -= 1
            else:
                result.append([l, r])
                l += 1
                r -= 1
                while l < r and nums[l] == nums[l - 1]:
                    l += 1
                while l < r and nums[r] == nums[r + 1]:
                    r -= 1

        return result

    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums = sorted(nums)

        result = []

        for i in range(0, len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j, k in self.twoSumSorted(nums, i + 1, -nums[i]):
                result.append([nums[i], nums[j], nums[k]])

        return result
